File: ModalLogic.f90
! モーダル論理の公理系（ML-MOD）をFORTRANで実装
! 圏論・代数的解釈（圏、群、環、体）を反映
! 公理と推論規則を関数として定義し、可能世界フレームをシミュレーション
PROGRAM ModalLogic
  IMPLICIT NONE

  ! モジュール定義：モーダル論理の構造と関数
  MODULE ModalLogicModule
    IMPLICIT NONE
    INTEGER, PARAMETER :: MAX_WORLDS = 100 ! 最大可能世界数
    INTEGER, PARAMETER :: MAX_PROPOSITIONS = 50 ! 最大命題数

    ! 可能世界フレームの構造体
    TYPE :: WorldFrame
      INTEGER :: num_worlds ! 世界の数
      LOGICAL :: relations(MAX_WORLDS, MAX_WORLDS) ! アクセス可能性関係 R(w,v)
      LOGICAL :: propositions(MAX_WORLDS, MAX_PROPOSITIONS) ! 各世界での命題の真偽
    END TYPE WorldFrame

    ! モーダル演算子の定義
    TYPE :: ModalOperator
      CHARACTER(LEN=10) :: name ! 演算子名（例："Box", "Diamond", "K_i", "O"）
    END TYPE ModalOperator

  CONTAINS
    ! 公理1：命題論理のトートロジー（Tautology-MOD）
    LOGICAL FUNCTION Tautology(A, B, C)
      LOGICAL, INTENT(IN) :: A, B, C
      ! 例：A → (B → A) のトートロジーをチェック
      Tautology = (.NOT. A .OR. (.NOT. B .OR. A))
      ! 圏論的解釈：ブール代数の恒等射
    END FUNCTION Tautology

    ! 公理2：分配公理（K-MOD）
    LOGICAL FUNCTION K_Axiom(frame, prop_A, prop_B, op_box)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:), prop_B(:) ! 命題A, Bの真偽（各世界）
      TYPE(ModalOperator), INTENT(IN) :: op_box
      LOGICAL :: box_A(MAX_WORLDS), box_B(MAX_WORLDS), box_imp(MAX_WORLDS)
      INTEGER :: i, j
      ! □(A → B) → (□A → □B)
      DO i = 1, frame%num_worlds
        box_A(i) = .TRUE.
        box_B(i) = .TRUE.
        box_imp(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            box_A(i) = box_A(i) .AND. prop_A(j)
            box_B(i) = box_B(i) .AND. prop_B(j)
            box_imp(i) = box_imp(i) .AND. (.NOT. prop_A(j) .OR. prop_B(j))
          END IF
        END DO
      END DO
      K_Axiom = ALL(.NOT. box_imp .OR. (.NOT. box_A .OR. box_B))
      ! 圏論的解釈：□は前像関手、代数的解釈：環Rの準同型
    END FUNCTION K_Axiom

    ! 公理3：双対公理（Dual-MOD）
    LOGICAL FUNCTION Dual_Axiom(frame, prop_A, op_box, op_diamond)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_box, op_diamond
      LOGICAL :: box_neg_A(MAX_WORLDS), diamond_A(MAX_WORLDS)
      INTEGER :: i, j
      ! ◇A ↔ ¬□¬A
      DO i = 1, frame%num_worlds
        box_neg_A(i) = .TRUE.
        diamond_A(i) = .FALSE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            box_neg_A(i) = box_neg_A(i) .AND. (.NOT. prop_A(j))
            diamond_A(i) = diamond_A(i) .OR. prop_A(j)
          END IF
        END DO
      END DO
      Dual_Axiom = ALL(diamond_A .EQV. (.NOT. box_neg_A))
      ! 圏論的解釈：◇は□の余関手
    END FUNCTION Dual_Axiom

    ! 公理4：反射公理（T-MOD）
    LOGICAL FUNCTION T_Axiom(frame, prop_A, op_box)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_box
      LOGICAL :: box_A(MAX_WORLDS)
      INTEGER :: i, j
      ! □A → A
      DO i = 1, frame%num_worlds
        box_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            box_A(i) = box_A(i) .AND. prop_A(j)
          END IF
        END DO
      END DO
      T_Axiom = ALL(.NOT. box_A .OR. prop_A)
      ! 圏論的解釈：反射的圏、群論的解釈：自明な群作用
    END FUNCTION T_Axiom

    ! 公理5：シリアル公理（D-MOD）
    LOGICAL FUNCTION D_Axiom(frame, prop_A, op_box, op_diamond)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_box, op_diamond
      LOGICAL :: box_A(MAX_WORLDS), diamond_A(MAX_WORLDS)
      INTEGER :: i, j
      ! □A → ◇A
      DO i = 1, frame%num_worlds
        box_A(i) = .TRUE.
        diamond_A(i) = .FALSE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            box_A(i) = box_A(i) .AND. prop_A(j)
            diamond_A(i) = diamond_A(i) .OR. prop_A(j)
          END IF
        END DO
      END DO
      D_Axiom = ALL(.NOT. box_A .OR. diamond_A)
      ! 圏論的解釈：シリアル圏、体論的解釈：非零ベクトル
    END FUNCTION D_Axiom

    ! 公理6：推移公理（4-MOD）
    LOGICAL FUNCTION Four_Axiom(frame, prop_A, op_box)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_box
      LOGICAL :: box_A(MAX_WORLDS), box_box_A(MAX_WORLDS)
      INTEGER :: i, j, k
      ! □A → □□A
      DO i = 1, frame%num_worlds
        box_A(i) = .TRUE.
        box_box_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            box_A(i) = box_A(i) .AND. prop_A(j)
            LOGICAL :: temp_box = .TRUE.
            DO k = 1, frame%num_worlds
              IF (frame%relations(j,k)) THEN
                temp_box = temp_box .AND. prop_A(k)
              END IF
            END DO
            box_box_A(i) = box_box_A(i) .AND. temp_box
          END IF
        END DO
      END DO
      Four_Axiom = ALL(.NOT. box_A .OR. box_box_A)
      ! 圏論的解釈：推移的圏、群論的解釈：推移的軌道
    END FUNCTION Four_Axiom

    ! 公理7：ユークリッド公理（5-MOD）
    LOGICAL FUNCTION Five_Axiom(frame, prop_A, op_box, op_diamond)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_box, op_diamond
      LOGICAL :: diamond_A(MAX_WORLDS), box_diamond_A(MAX_WORLDS)
      INTEGER :: i, j, k
      ! ◇A → □◇A
      DO i = 1, frame%num_worlds
        diamond_A(i) = .FALSE.
        box_diamond_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            diamond_A(i) = diamond_A(i) .OR. prop_A(j)
            LOGICAL :: temp_diamond = .FALSE.
            DO k = 1, frame%num_worlds
              IF (frame%relations(j,k)) THEN
                temp_diamond = temp_diamond .OR. prop_A(k)
              END IF
            END DO
            box_diamond_A(i) = box_diamond_A(i) .AND. temp_diamond
          END IF
        END DO
      END DO
      Five_Axiom = ALL(.NOT. diamond_A .OR. box_diamond_A)
      ! 圏論的解釈：ユークリッド圏、体論的解釈：ガロア拡大
    END FUNCTION Five_Axiom

    ! 公理8：対称公理（B-MOD）
    LOGICAL FUNCTION B_Axiom(frame, prop_A, op_box, op_diamond)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_box, op_diamond
      LOGICAL :: box_diamond_A(MAX_WORLDS)
      INTEGER :: i, j, k
      ! A → □◇A
      DO i = 1, frame%num_worlds
        box_diamond_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            LOGICAL :: temp_diamond = .FALSE.
            DO k = 1, frame%num_worlds
              IF (frame%relations(j,k)) THEN
                temp_diamond = temp_diamond .OR. prop_A(k)
              END IF
            END DO
            box_diamond_A(i) = box_diamond_A(i) .AND. temp_diamond
          END IF
        END DO
      END DO
      B_Axiom = ALL(.NOT. prop_A .OR. box_diamond_A)
      ! 圏論的解釈：対称的圏、群論的解釈：アーベル群
    END FUNCTION B_Axiom

    ! 知識論理の公理（例）
    ! 公理9：知識の真公理（KE-MOD）
    LOGICAL FUNCTION KE_Axiom(frame, prop_A, op_k)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_k
      LOGICAL :: k_A(MAX_WORLDS)
      INTEGER :: i, j
      ! K_i A → A
      DO i = 1, frame%num_worlds
        k_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            k_A(i) = k_A(i) .AND. prop_A(j)
          END IF
        END DO
      END DO
      KE_Axiom = ALL(.NOT. k_A .OR. prop_A)
      ! 圏論的解釈：反射的知識圏
    END FUNCTION KE_Axiom

    ! 公理10：正の内省公理（K4-MOD）
    LOGICAL FUNCTION K4_Axiom(frame, prop_A, op_k)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_k
      LOGICAL :: k_A(MAX_WORLDS), k_k_A(MAX_WORLDS)
      INTEGER :: i, j, k
      ! K_i A → K_i K_i A
      DO i = 1, frame%num_worlds
        k_A(i) = .TRUE.
        k_k_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            k_A(i) = k_A(i) .AND. prop_A(j)
            LOGICAL :: temp_k = .TRUE.
            DO k = 1, frame%num_worlds
              IF (frame%relations(j,k)) THEN
                temp_k = temp_k .AND. prop_A(k)
              END IF
            END DO
            k_k_A(i) = k_k_A(i) .AND. temp_k
          END IF
        END DO
      END DO
      K4_Axiom = ALL(.NOT. k_A .OR. k_k_A)
      ! 圏論的解釈：推移的知識圏
    END FUNCTION K4_Axiom

    ! 公理11：負の内省公理（K5-MOD）
    LOGICAL FUNCTION K5_Axiom(frame, prop_A, op_k)
      TYPE(WorldFrame), INTENT(IN) :: frame
      LOGICAL, INTENT(IN) :: prop_A(:)
      TYPE(ModalOperator), INTENT(IN) :: op_k
      LOGICAL :: not_k_A(MAX_WORLDS), k_not_k_A(MAX_WORLDS)
      INTEGER :: i, j, k
      ! ¬K_i A → K_i ¬K_i A
      DO i = 1, frame%num_worlds
        not_k_A(i) = .FALSE.
        k_not_k_A(i) = .TRUE.
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            not_k_A(i) = not_k_A(i) .OR. (.NOT. prop_A(j))
            LOGICAL :: temp_not_k = .FALSE.
            DO k = 1, frame%num_worlds
              IF (frame%relations(j,k)) THEN
                temp_not_k = temp_not_k .OR. (.NOT. prop_A(k))
              END IF
            END DO
            k_not_k_A(i) = k_not_k_A(i) .AND. temp_not_k
          END IF
        END DO
      END DO
      K5_Axiom = ALL(.NOT. not_k_A .OR. k_not_k_A)
      ! 圏論的解釈：ユークリッド知識圏、体論的解釈：ガロア固定点
    END FUNCTION K5_Axiom

    ! フレーム条件のチェック（例：反射性）
    LOGICAL FUNCTION IsReflexive(frame)
      TYPE(WorldFrame), INTENT(IN) :: frame
      INTEGER :: i
      IsReflexive = .TRUE.
      DO i = 1, frame%num_worlds
        IF (.NOT. frame%relations(i,i)) THEN
          IsReflexive = .FALSE.
          EXIT
        END IF
      END DO
      ! 圏論的解釈：恒等射の存在
    END FUNCTION IsReflexive

    ! フレーム条件のチェック（例：推移性）
    LOGICAL FUNCTION IsTransitive(frame)
      TYPE(WorldFrame), INTENT(IN) :: frame
      INTEGER :: i, j, k
      IsTransitive = .TRUE.
      DO i = 1, frame%num_worlds
        DO j = 1, frame%num_worlds
          IF (frame%relations(i,j)) THEN
            DO k = 1, frame%num_worlds
              IF (frame%relations(j,k) .AND. .NOT. frame%relations(i,k)) THEN
                IsTransitive = .FALSE.
                RETURN
              END IF
            END DO
          END IF
        END DO
      END DO
      ! 圏論的解釈：射の合成、群論的解釈：推移的軌道
    END FUNCTION IsTransitive

  END MODULE ModalLogicModule

  ! メインルーチン
  USE ModalLogicModule
  TYPE(WorldFrame) :: frame
  TYPE(ModalOperator) :: op_box, op_diamond, op_k
  LOGICAL :: prop_A(MAX_WORLDS), prop_B(MAX_WORLDS)
  INTEGER :: i

  ! 可能世界フレームの初期化（例：3つの世界）
  frame%num_worlds = 3
  frame%relations = .FALSE.
  ! 例：S5フレーム（反射的、推移的、ユークリッド的）
  frame%relations(1,1) = .TRUE.
  frame%relations(1,2) = .TRUE.
  frame%relations(1,3) = .TRUE.
  frame%relations(2,1) = .TRUE.
  frame%relations(2,2) = .TRUE.
  frame%relations(2,3) = .TRUE.
  frame%relations(3,1) = .TRUE.
  frame%relations(3,2) = .TRUE.
  frame%relati
ons(3,3) = .TRUE.

  ! 命題の初期化
  prop_A = [.TRUE., .FALSE., .TRUE.] ! 世界1,3でAが真
  prop_B = [.FALSE., .TRUE., .TRUE.] ! 世界2,3でBが真

  ! モーダル演算子の初期化
  op_box%name = "Box"
  op_diamond%name = "Diamond"
  op_k%name = "K_i"

  ! 公理のテスト
  PRINT *, "Testing Tautology-MOD: ", Tautology(.TRUE., .FALSE., .TRUE.)
  PRINT *, "Testing K-MOD: ", K_Axiom(frame, prop_A, prop_B, op_box)
  PRINT *, "Testing Dual-MOD: ", Dual_Axiom(frame, prop_A, op_box, op_diamond)
  PRINT *, "Testing T-MOD: ", T_Axiom(frame, prop_A, op_box)
  PRINT *, "Testing D-MOD: ", D_Axiom(frame, prop_A, op_box, op_diamond)
  PRINT *, "Testing 4-MOD: ", Four_Axiom(frame, prop_A, op_box)
  PRINT *, "Testing 5-MOD: ", Five_Axiom(frame, prop_A, op_box, op_diamond)
  PRINT *, "Testing B-MOD: ", B_Axiom(frame, prop_A, op_box, op_diamond)
  PRINT *, "Testing KE-MOD: ", KE_Axiom(frame, prop_A, op_k)
  PRINT *, "Testing K4-MOD: ", K4_Axiom(frame, prop_A, op_k)
  PRINT *, "Testing K5-MOD: ", K5_Axiom(frame, prop_A, op_k)
  PRINT *, "Frame Reflexive: ", IsReflexive(frame)
  PRINT *, "Frame Transitive: ", IsTransitive(frame)

END PROGRAM ModalLogic

File: ModalLogic.jl
# モーダル論理の公理系（ML-MOD）をJuliaで実装
# 圏論・代数的解釈（圏、群、環、体）を反映
# 公理と推論規則を関数として定義し、可能世界フレームをシミュレーション

module ModalLogic

# 定数
const MAX_WORLDS = 100
const MAX_PROPOSITIONS = 50

# 可能世界フレームの構造体
mutable struct WorldFrame
    num_worlds::Int
    relations::Matrix{Bool} # アクセス可能性関係 R(w,v)
    propositions::Matrix{Bool} # 各世界での命題の真偽
end

# モーダル演算子の構造体
struct ModalOperator
    name::String # 演算子名（例："Box", "Diamond", "K_i", "O"）
end

# 公理1：命題論理のトートロジー（Tautology-MOD）
function tautology(A::Bool, B::Bool, C::Bool)::Bool
    # 例：A → (B → A) のトートロジーをチェック
    return !A || (!B || A)
    # 圏論的解釈：ブール代数の恒等射
end

# 公理2：分配公理（K-MOD）
function k_axiom(frame::WorldFrame, prop_A::Vector{Bool}, prop_B::Vector{Bool}, op_box::ModalOperator)::Bool
    # □(A → B) → (□A → □B)
    box_A = fill(true, frame.num_worlds)
    box_B = fill(true, frame.num_worlds)
    box_imp = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                box_A[i] = box_A[i] && prop_A[j]
                box_B[i] = box_B[i] && prop_B[j]
 reigns                box_imp[i] = box_imp[i] && (!prop_A[j] || prop_B[j])
            end
        end
    end
    return all(!box_imp[i] || (!box_A[i] || box_B[i]) for i in 1:frame.num_worlds)
    # 圏論的解釈：□は前像関手、代数的解釈：環Rの準同型
end

# 公理3：双対公理（Dual-MOD）
function dual_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_box::ModalOperator, op_diamond::ModalOperator)::Bool
    # ◇A ↔ ¬□¬A
    box_neg_A = fill(true, frame.num_worlds)
    diamond_A = fill(false, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                box_neg_A[i] = box_neg_A[i] && !prop_A[j]
                diamond_A[i] = diamond_A[i] || prop_A[j]
            end
        end
    end
    return all(diamond_A[i] == !box_neg_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：◇は□の余関手
end

# 公理4：反射公理（T-MOD）
function t_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_box::ModalOperator)::Bool
    # □A → A
    box_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                box_A[i] = box_A[i] && prop_A[j]
            end
        end
    end
    return all(!box_A[i] || prop_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：反射的圏、群論的解釈：自明な群作用
end

# 公理5：シリアル公理（D-MOD）
function d_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_box::ModalOperator, op_diamond::ModalOperator)::Bool
    # □A → ◇A
    box_A = fill(true, frame.num_worlds)
    diamond_A = fill(false, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                box_A[i] = box_A[i] && prop_A[j]
                diamond_A[i] = diamond_A[i] || prop_A[j]
            end
        end
    end
    return all(!box_A[i] || diamond_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：シリアル圏、体論的解釈：非零ベクトル
end

# 公理6：推移公理（4-MOD）
function four_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_box::ModalOperator)::Bool
    # □A → □□A
    box_A = fill(true, frame.num_worlds)
    box_box_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                box_A[i] = box_A[i] && prop_A[j]
                temp_box = true
                for k in 1:frame.num_worlds
                    if frame.relations[j,k]
                        temp_box = temp_box && prop_A[k]
                    end
                end
                box_box_A[i] = box_box_A[i] && temp_box
            end
        end
    end
    return all(!box_A[i] || box_box_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：推移的圏、群論的解釈：推移的軌道
end

# 公理7：ユークリッド公理（5-MOD）
function five_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_box::ModalOperator, op_diamond::ModalOperator)::Bool
    # ◇A → □◇A
    diamond_A = fill(false, frame.num_worlds)
    box_diamond_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                diamond_A[i] = diamond_A[i] || prop_A[j]
                temp_diamond = false
                for k in 1:frame.num_worlds
                    if frame.relations[j,k]
                        temp_diamond = temp_diamond || prop_A[k]
                    end
                end
                box_diamond_A[i] = box_diamond_A[i] && temp_diamond
            end
        end
    end
    return all(!diamond_A[i] || box_diamond_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：ユークリッド圏、体論的解釈：ガロア拡大
end

# 公理8：対称公理（B-MOD）
function b_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_box::ModalOperator, op_diamond::ModalOperator)::Bool
    # A → □◇A
    box_diamond_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                temp_diamond = false
                for k in 1:frame.num_worlds
                    if frame.relations[j,k]
                        temp_diamond = temp_diamond || prop_A[k]
                    end
                end
                box_diamond_A[i] = box_diamond_A[i] && temp_diamond
            end
        end
    end
    return all(!prop_A[i] || box_diamond_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：対称的圏、群論的解釈：アーベル群
end

# 知識論理の公理
# 公理9：知識の真公理（KE-MOD）
function ke_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_k::ModalOperator)::Bool
    # K_i A → A
    k_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                k_A[i] = k_A[i] && prop_A[j]
            end
        end
    end
    return all(!k_A[i] || prop_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：反射的知識圏
end

# 公理10：正の内省公理（K4-MOD）
function k4_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_k::ModalOperator)::Bool
    # K_i A → K_i K_i A
    k_A = fill(true, frame.num_worlds)
    k_k_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                k_A[i] = k_A[i] && prop_A[j]
                temp_k = true
                for k in 1:frame.num_worlds
                    if frame.relations[j,k]
                        temp_k = temp_k && prop_A[k]
                    end
                end
                k_k_A[i] = k_k_A[i] && temp_k
            end
        end
    end
    return all(!k_A[i] || k_k_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：推移的知識圏
end

# 公理11：負の内省公理（K5-MOD）
function k5_axiom(frame::WorldFrame, prop_A::Vector{Bool}, op_k::ModalOperator)::Bool
    # ¬K_i A → K_i ¬K_i A
    not_k_A = fill(false, frame.num_worlds)
    k_not_k_A = fill(true, frame.num_worlds)
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                not_k_A[i] = not_k_A[i] || !prop_A[j]
                temp_not_k = false
                for k in 1:frame.num_worlds
                    if frame.relations[j,k]
                        temp_not_k = temp_not_k || !prop_A[k]
                    end
                end
                k_not_k_A[i] = k_not_k_A[i] && temp_not_k
            end
        end
    end
    return all(!not_k_A[i] || k_not_k_A[i] for i in 1:frame.num_worlds)
    # 圏論的解釈：ユークリッド知識圏、体論的解釈：ガロア固定点
end

# フレーム条件：反射性
function is_reflexive(frame::WorldFrame)::Bool
    for i in 1:frame.num_worlds
        if !frame.relations[i,i]
            return false
        end
    end
    return true
    # 圏論的解釈：恒等射の存在
end

# フレーム条件：推移性
function is_transitive(frame::WorldFrame)::Bool
    for i in 1:frame.num_worlds
        for j in 1:frame.num_worlds
            if frame.relations[i,j]
                for k in 1:frame.num_worlds
                    if frame.relations[j,k] && !frame.relations[i,k]
                        return false
                    end
                end
            end
        end
    end
    return true
    # 圏論的解釈：射の合成、群論的解釈：推移的軌道
end

export WorldFrame, ModalOperator, tautology, k_axiom, dual_axiom, t_axiom, d_axiom,
       four_axiom, five_axiom, b_axiom, ke_axiom, k4_axiom, k5_axiom,
       is_reflexive, is_transitive

end # module ModalLogic

# メインスクリプト
using .ModalLogic

# 可能世界フレームの初期化（例：3つの世界、S5フレーム）
num_worlds = 3
relations = fill(true, num_worlds, num_worlds) # 全アクセス可能（S5）
propositions = falses(num_worlds, MAX_PROPOSITIONS)
frame = WorldFrame(num_worlds, relations, propositions)

# 命題の初期化
prop_A = [true, false, true] # 世界1,3でAが真
prop_B = [false, true, true] # 世界2,3でBが真

# モーダル演算子の初期化
op_box = ModalOperator("Box")
op_diamond = ModalOperator("Diamond")
op_k = ModalOperator("K_i")

# 公理とフレーム条件のテスト
println("Testing Tautology-MOD: ", tautology(true, false, true))
println("Testing K-MOD: ", k_axiom(frame, prop_A, prop_B, op_box))
println("Testing Dual-MOD: ", dual_axiom(frame, prop_A, op_box, op_diamond))
println("Testing T-MOD: ", t_axiom(frame, prop_A, op_box))
println("Testing D-MOD: ", d_axiom(frame, prop_A, op_box, op_diamond))
println("Testing 4-MOD: ", four_axiom(frame, prop_A, op_box))
println("Testing 5-MOD: ", five_axiom(frame, prop_A, op_box, op_diamond))
println("Testing B-MOD: ", b_axiom(frame, prop_A, op_box, op_diamond))
println("Testing KE-MOD: ", ke_axiom(frame, prop_A, op_k))
println("Testing K4-MOD: ", k4_axiom(frame, prop_A, op_k))
println("Testing K5-MOD: ", k5_axiom(frame, prop_A, op_k))
println("Frame Reflexive: ", is_reflexive(frame))
println("Frame Transitive: ", is_transitive(frame))
